include/osmium/io/metadata_options.hpp

```diff
@@ -0,0 +1,204 @@
+                    }
+                }
+                m_options = static_cast<options>(opts);
+            }
+
+            metadata_options(const osmium::OSMObject& object) {
+                int opts = 0;
+                if (object.version() > 0) {
+                    opts |= options::md_version;
+                }
+                if (object.timestamp().valid()) {
+                    opts |= options::md_timestamp;
+                }
+                if (object.changeset() > 0) {
+                    opts |= options::md_changeset;
+                }
+                // Objects by anonymous users don't have these attributes set. There is no way
+                // to distinguish them from objects with a reduced number of metadata fields.
+                if (object.uid() > 0) {
+                    opts |= options::md_uid;
+                }
+                if (object.user() && strlen(object.user()) > 0) {
+                    opts |= options::md_user;
+                }
+                m_options = static_cast<options>(opts);
+            }
+
+            /// At least one metadata attribute should be stored.
+            bool any() const noexcept {
+                return m_options != 0;
+            }
+
+            /// All metadata attributes should be stored.
+            bool all() const noexcept {
+                return m_options == options::md_all;
+            }
+
+            /// No metadata attributes should be stored.
+            bool none() const noexcept {
+                return m_options == 0;
+            }
+
+            bool version() const noexcept {
+                return (m_options & options::md_version) != 0;
+            }
+
+            bool timestamp() const noexcept {
+                return (m_options & options::md_timestamp) != 0;
+            }
+
+            bool changeset() const noexcept {
+                return (m_options & options::md_changeset) != 0;
+            }
+
+            bool uid() const noexcept {
+                return (m_options & options::md_uid) != 0;
+            }
+
+            bool user() const noexcept {
+                return (m_options & options::md_user) != 0;
+            }
+
+            metadata_options operator&=(const metadata_options& other) {
+                m_options = static_cast<options>(other.m_options & m_options);
+                return *this;
+            }
+
+            metadata_options operator|=(const metadata_options& other) {
+                m_options = static_cast<options>(other.m_options | m_options);
+                return *this;
+            }
+
+        }; // class metadata_options
+
+        template <typename TChar, typename TTraits>
+        inline bool write_option(std::basic_ostream<TChar, TTraits>& out, bool already_written, const char* name) {
+            if (already_written) {
+                out << '+';
+            }
+            out << name;
+            return true;
+        }
+
+        template <typename TChar, typename TTraits>
```

Comment:
I suggest writing a `metadata_options_to_string` function that returns a std::string. There you can assemble the string by always appending the metadata option name and the `+` sign, last thing then is to `s.resize(s.size() - 1)` to remove the last `+`. Makes the code much simpler and can be useful if you need a string that you don't want to output to a stream. The ostream operator is a one-liner then.

The tests you have below are much simpler to write, too.

Reply:
You mean that [method](https://github.com/osmcode/libosmium/pull/248/commits/31e34f10c52f08a53469e06e98867982bf31560a#diff-195c657e6eb2642212801e6799d5dd78R159) which I removed in my last commit?

Reply:
Yep. That's it. :-)

Also one tip there: If you move the `std::string result` to the beginning and set it even in the `all` and `none` case, the same `std::string` is always used as the return value which should allow return value optimization to work which would remove a copy.